L1Trigger/CSCTriggerPrimitives/src/CSCUpgradeMotherboardLUTGenerator.cc

```diff
@@ -0,0 +1,356 @@
+{
+  bool rpcGeometryAvailable(false);
+  if (rpc_g != nullptr) {
+    LogTrace("CSCUpgradeMotherboardLUTGenerator")<< "+++ generateLUTsME3141() called for ME3141 chamber! +++ \n";
+    rpcGeometryAvailable = true;
+  }
+  
+  if (not rpcGeometryAvailable){
+    LogTrace("CSCUpgradeMotherboardLUTGenerator") << "+++ generateLUTsME3141() called for ME3141 chamber without valid RPC geometry! +++ \n";
+    return;
+  }
+
+  // CSC trigger geometry
+  const int chid = CSCTriggerNumbering::chamberFromTriggerLabels(theSector, theSubsector, theStation, theTrigChamber);
+  const CSCDetId csc_id(theEndcap, theStation, 1, chid, 0);
+  const CSCChamber* cscChamber(csc_g->chamber(csc_id));
+  const CSCLayer* keyLayer(cscChamber->layer(3));
+
+  // RPC trigger geometry
+  const RPCDetId rpc_id(getRPCfromCSC(csc_id));
+  const RPCChamber* rpcChamber(rpc_g->chamber(rpc_id));
+  const RPCRoll* randRoll(rpcChamber->roll(2));
+
+  // LUTs
+  std::vector<std::pair<double,double> > rpcRollToEtaLimits_ = rpcRollToEtaLimitsLUT(rpcChamber);
+  std::vector<std::pair<double, double> > cscWGToEtaLimits_ = cscWgToEtaLimitsLUT(keyLayer);
+  std::vector<std::pair<int,int> > cscWgToRpcRoll_ = cscWgToRollLUT(cscWGToEtaLimits_, rpcRollToEtaLimits_);
+  std::vector<std::pair<int,int> > cscHsToRpcStrip_ = cscHsToRpcStripLUT(keyLayer,randRoll,5,155);
+  std::vector<int> rpcStripToCscHs_ = rpcStripToCscHsLUT(keyLayer, randRoll);
+
+  std::stringstream os;
+  os << "ME31/41 "<< csc_id <<std::endl;
+
+  os << "RPC roll to eta limits" << std::endl;
+  os << rpcRollToEtaLimits_;
+
+  os << "WG to eta limits" << std::endl;
+  os << cscWGToEtaLimits_;
+
+  os << "WG to Roll" << std::endl;
+  os << cscWgToRpcRoll_;
+
+  os << "CSC HS to RPC strip LUT in ME3141" << std::endl;
+  os << cscHsToRpcStrip_;
+  
+  os << "RPC strip to CSC HS LUT in ME3141" << std::endl;
+  os << rpcStripToCscHs_;
+
+  // print LUTs
+  LogTrace("CSCUpgradeMotherboardLUTGenerator") << os.str();
+}
+
+int CSCUpgradeMotherboardLUTGenerator::assignRoll(const std::vector<std::pair<double,double> >& lut_, double eta) const
+{
+  int result = -99;
+  for(const auto& p : lut_) {
+    const float minEta(p.first);
+    const float maxEta(p.second);
+    if (minEta <= std::abs(eta) and std::abs(eta) < maxEta) {
+      result = p.first;
+      break;
+    }
+  }
+  return result;
+}
+
+std::vector<std::pair<double,double> >
+CSCUpgradeMotherboardLUTGenerator::gemRollToEtaLimitsLUT(const GEMChamber* gemChamber) const
+{
+  std::vector<std::pair<double,double> > lut;
+  for(const auto& roll : gemChamber->etaPartitions()) {
+    const float half_striplength(roll->specs()->specificTopology().stripLength()/2.);
+    const LocalPoint lp_top(0., half_striplength, 0.);
+    const LocalPoint lp_bottom(0., -half_striplength, 0.);
+    const GlobalPoint gp_top(roll->toGlobal(lp_top));
+    const GlobalPoint gp_bottom(roll->toGlobal(lp_bottom));
+    lut.push_back(std::make_pair(std::abs(gp_top.eta()), std::abs(gp_bottom.eta())));
```

Comment:
The GEM geometry is symmetric with respect to eta. At least to first order. There may be small effects by matching the simulation to an imperfect installation. If it would be necessary to have LUTs for each endcap, that can be taken care of later.